packages/ts/react-grid/src/header-filter.tsx

```diff
@@ -0,0 +1,81 @@
+import { Item } from '@hilla/react-components/Item.js';
+import { ListBox } from '@hilla/react-components/ListBox.js';
+import { Select, type SelectElement } from '@hilla/react-components/Select.js';
+import { TextField, type TextFieldElement } from '@hilla/react-components/TextField.js';
+import { useContext, useEffect, useRef, useState, type ReactElement } from 'react';
+import { HeaderColumnContext } from './header-column-context.js';
+import css from './header-filter.module.css';
+import Matcher from './types/dev/hilla/crud/filter/PropertyStringFilter/Matcher';
+
+export function HeaderFilter(): ReactElement {
+  const context = useContext(HeaderColumnContext)!;
+  const [matcher, setMatcher] = useState(Matcher.GREATER_THAN);
+  const [filterValue, setFilterValue] = useState('');
+  const select = useRef<SelectElement>(null);
+  useEffect(() => {
+    // Workaround for https://github.com/vaadin/react-components/issues/148
+    setTimeout(() => {
+      if (select.current) {
+        select.current.requestContentUpdate();
+      }
+    }, 1);
+  }, []);
+
+  useEffect(() => {
```

Comment:
Nit: I'd suggest to just use a function with arguments for value + matcher, and call that from the event handlers instead of relying on detecting state changes. It's a bit unexpected that you would get a `setPropertyFilter` call just from rendering the component initially. That way you can also remove `changed` checks in the filter implementation.

Reply:
Where would you track the current values?

Reply:
I would still keep the states, but set them from said function:
```ts
const [matcher, setMatcher] = useState(Matcher.GREATER_THAN);
const [filterValue, setFilterValue] = useState('');

function updateFilter(newFilterValue: string, newMatcher: string) {
  setFilterValue(newFilterValue);
  setMatcher(newMatcher);

  const filter = {
    propertyId: context.propertyInfo.name,
    filterValue,
    matcher,
  };
  // eslint-disable-next-line
  (filter as any).t = 'propertyString';
  context.setPropertyFilter(filter);
} 
```

In cases where you don't want to change one of the states, for example in the select, just pass the current state as argument.

Reply:
Ok, what I thought. It's a bit fiddly because you cannot use both values from the state as they don't update immediately. But probably works better this way.

Fixed

Reply:
Sorry, I meant something like this:
```diff
@@ -9,18 +9,6 @@
 import Matcher from './types/dev/hilla/crud/filter/PropertyStringFilter/Matcher';
 import { PropertyInfo } from './utils.js';
 
-function triggerFilterUpdate(matcher: Matcher, filterValue: string, context: HeaderColumnContext) {
-  // Update filter
-  const filter = {
-    propertyId: context.propertyInfo.name,
-    filterValue,
-    matcher,
-  };
-  // eslint-disable-next-line
-  (filter as any).t = 'propertyString';
-  context.setPropertyFilter(filter);
-}
-
 export function HeaderFilter(): ReactElement {
   const context = useContext(HeaderColumnContext)!;
   const [matcher, setMatcher] = useState(Matcher.GREATER_THAN);
@@ -35,13 +23,26 @@
     }, 1);
   }, []);
 
+  function updateFilter(newFilterValue: string, newMatcher: Matcher) {
+    setFilterValue(newFilterValue);
+    setMatcher(newMatcher);
+    const filter = {
+      propertyId: context.propertyInfo.name,
+      filterValue: newFilterValue,
+      matcher: newMatcher,
+    };
+    // eslint-disable-next-line
+    (filter as any).t = 'propertyString';
+    context.setPropertyFilter(filter);
+  }
+
   if (context.propertyInfo.modelType === 'string') {
     return (
       <TextField
         placeholder="Filter..."
         onInput={(e: any) => {
           const fieldValue = ((e as InputEvent).target as TextFieldElement).value;
-          triggerFilterUpdate(Matcher.CONTAINS, fieldValue, context);
+          updateFilter(fieldValue, Matcher.CONTAINS);
         }}
       ></TextField>
     );
@@ -52,8 +53,7 @@
           ref={select}
           onValueChanged={(e) => {
             const newMatcher = e.detail.value as Matcher;
-            setMatcher(newMatcher);
-            triggerFilterUpdate(newMatcher, filterValue, context);
+            updateFilter(filterValue, newMatcher);
           }}
           renderer={() => (
             <ListBox>
@@ -75,8 +75,7 @@
           placeholder="Filter..."
           onInput={(e) => {
             const fieldValue = ((e as InputEvent).target as TextFieldElement).value;
-            setFilterValue(fieldValue);
-            triggerFilterUpdate(matcher, fieldValue, context);
+            updateFilter(fieldValue, matcher);
           }}
         />
       </>
```

Reply:
That's another way of writing the same thing :)

Reply:
A bit cleaner, applied